### PR TITLE
Hy/plpython3

### PR DESCRIPTION
**Pull request by yihong0618**
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

--- 

add plpython3 path and guc warning

### CODE REVIEW

File: src/pl/plpython/plpy_main.c [hunk 1]
```diff
@@ -80,6 +82,24 @@ cancel_pending_hook_type prev_cancel_pending_hook;
 void PLy_handle_cancel_interrupt(void);
 
 bool PLy_enter_python_intepreter = false;
+bool PLy_path_added = false;
+
+/* GUC variables */
+#if PY_MAJOR_VERSION >= 3
+char *plpython3_path = NULL;
```
Comment (beeender) [inline src/pl/plpython/plpy_main.c]: make it static, it should not be used out of this file.

```
static char *plpython3_path = NULL;
```

File: src/pl/plpython/plpy_main.c [hunk 2]
```diff
@@ -80,6 +82,24 @@ cancel_pending_hook_type prev_cancel_pending_hook;
...
+
+bool 
+plpython3_check_python_path(char **newval, void **extra, GucSource source) {
+	if (PLy_path_added)
```
Comment (beeender) [inline src/pl/plpython/plpy_main.c]: I still don't understand why cannot we use `inited`.

Even user sets it twice, as long as `PLy_initialize` has not been executed, it doesn't take effect yet.

File: src/pl/plpython/plpy_main.c [hunk 3]
```diff
@@ -80,6 +82,24 @@ cancel_pending_hook_type prev_cancel_pending_hook;
...
+	{
+		GUC_check_errmsg("SET PYTHONPATH for plpython3 can only set once in one session");
```
Comment (beeender) [inline src/pl/plpython/plpy_main.c]: If the above can be changed, then we need to modify this message.

File: src/pl/plpython/plpy_main.c [hunk 4]
```diff
@@ -80,6 +82,24 @@ cancel_pending_hook_type prev_cancel_pending_hook;
...
+		return false;
+	}
+	if (strcmp(*newval, "") != 0)
```
Comment (beeender) [inline src/pl/plpython/plpy_main.c]: if we have this check, then the `PLy_path_added` is not needed. Since we can always check if the `plpython3_path` is `NULL`. Right?

File: src/pl/plpython/plpy_main.c [hunk 5]
```diff
@@ -137,6 +157,18 @@ _PG_init(void)
 	prev_cancel_pending_hook = cancel_pending_hook;
 	cancel_pending_hook = PLy_handle_cancel_interrupt;
 
+#if PY_MAJOR_VERSION >= 3
+	DefineCustomStringVariable("plpython3.python_path",
```
Comment (beeender) [inline src/pl/plpython/plpy_main.c:160]: What does `u` mean in the `plpython3u`? I am thinking about shall we align the GUC name with the pl name `plpython3u` like `plpython3u.python_path` here?

File: src/pl/plpython/plpy_main.c [hunk 6]
```diff
@@ -162,18 +194,23 @@ PLy_initialize(void)
...
+	/* PYTHONPATH and PYTHONHOME has been set to GPDB's python2.7 in Postmaster when
+	 * gpstart. So for plpython3u, we need to unset PYTHONPATH and PYTHONHOME.
+	 * if user set PYTHONPATH then we set it in the env
+	 */
+	if (plpython3_path && *plpython3_path) 
+		setenv("PYTHONPATH", plpython3_path, 1);
+	else 
+		unsetenv("PYTHONPATH");
+#endif
+	unsetenv("PYTHONHOME");
```
Comment (beeender) [inline src/pl/plpython/plpy_main.c]: I am not quite sure if there is anyone relies on the `PYTHONHOME` for plpython2 for now. Probably it will be safer to have this for python3 only.

File: src/pl/plpython/sql/plpython3_path_guc.sql [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+set plpython3.python_path='/bar';
+
+-- Set up test path functions first.
+CREATE FUNCTION test_path_added() 
+RETURNS text AS $$
+  import sys
+  return str(sys.path[0])
+$$ language plpython3u;
+
+SELECT test_path_added();
```
Comment (beeender) [inline src/pl/plpython/sql/plpython3_path_guc.sql:17]: Add one more statement to set the GUC again here. And it is expected to fail.

File: src/pl/plpython/plpy_main.c (hunk 5 above)
Comment (yihong0618) [reply src/pl/plpython/plpy_main.c:160]: I think we should follow the `plperl` and 'plpgsql' logic which `guc` name has not `u`

File: src/pl/plpython/plpy_main.c (hunk 6 above)
Comment (yihong0618) [reply src/pl/plpython/plpy_main.c]: copy that.

File: src/pl/plpython/plpy_main.c (hunk 4 above)
Comment (yihong0618) [reply src/pl/plpython/plpy_main.c]: not for sure we need to make the default value empty string otherwise will cause the error (definecostumstring can not default NULL).

File: src/pl/plpython/plpy_main.c (hunk 2 above)
Comment (yihong0618) [reply src/pl/plpython/plpy_main.c]: because I think the logic is user may not create `plpython3u` language or call the function.
they just load `plython3u` then the `inited` can not access

File: src/pl/plpython/plpy_main.c (hunk 1 above)
Comment (yihong0618) [reply src/pl/plpython/plpy_main.c]: copy that.

File: src/pl/plpython/sql/plpython3_path_guc.sql (hunk 7 above)
Comment (yihong0618) [reply src/pl/plpython/sql/plpython3_path_guc.sql:17]: got it

File: src/pl/plpython/plpy_main.c [hunk 8]
```diff
@@ -84,19 +84,19 @@ void PLy_handle_cancel_interrupt(void);
...
-char *plpython3_path = NULL;
+static char *plpython3_path = NULL;
 
 bool 
 plpython3_check_python_path(char **newval, void **extra, GucSource source) {
-	if (PLy_path_added)
+	if (inited)
 	{
-		GUC_check_errmsg("SET PYTHONPATH for plpython3 can only set once in one session");
+		GUC_check_errmsg("SET PYTHONPATH faild, we can only set it before plpython3u init in one session");
```
Comment (beeender) [inline src/pl/plpython/plpy_main.c]: typo, `faild`, also maybe rephrase it to:

`The GUC value can only be changed before initializing the python interpreter.`

**Review by beeender — COMMENTED** [review]
looks good except the error message

File: src/pl/plpython/plpy_main.c [hunk 9]
```diff
@@ -80,6 +82,24 @@ cancel_pending_hook_type prev_cancel_pending_hook;
...
+bool PLy_path_added = false;
+
+static bool inited = false;
+
+/* GUC variables */
+#if PY_MAJOR_VERSION >= 3
+static char *plpython3_path = NULL;
+
+bool 
+plpython3_check_python_path(char **newval, void **extra, GucSource source) {
```
Comment (beeender) [inline src/pl/plpython/plpy_main.c:93]: and make this function static as well. No need to be exposed outside.

File: src/pl/plpython/expected/plpython3_path_guc.out [hunk 10]
```diff
@@ -0,0 +1,33 @@
+--
+-- Tests for functions that return set PYTHONPATH guc only for gpdb6 plpython3
+--
+LOAD 'plpython3';
+set plpython3.python_path='/foo';
```
Comment (beeender) [inline src/pl/plpython/expected/plpython3_path_guc.out:11]: you had `show plpython3.python_path` before, why did you remove it? I think it is reasonable to have it since it ensures the default value is empty.

Comment (yihong0618) [reply src/pl/plpython/expected/plpython3_path_guc.out:11]: because...it may set in the last time in our `ENV`

Comment (yihong0618) [reply src/pl/plpython/expected/plpython3_path_guc.out:11]: if it is a new `ENV` that is ok

File: src/pl/plpython/plpy_main.c [hunk 11]
```diff
@@ -137,6 +156,18 @@ _PG_init(void)
 	prev_cancel_pending_hook = cancel_pending_hook;
 	cancel_pending_hook = PLy_handle_cancel_interrupt;
 
+#if PY_MAJOR_VERSION >= 3
+	DefineCustomStringVariable("plpython3.python_path",
+							gettext_noop("Python path for plpython3."),
```
Comment (beeender) [inline src/pl/plpython/plpy_main.c]: last comment :
'PYTHONPATH' for plpython3 will be clearer IMO.

**Review by beeender — APPROVED** [review]

File: src/pl/plpython/plpy_main.c (hunk 11 above)
Comment (yihong0618) [reply src/pl/plpython/plpy_main.c]: copy that.

**Comment by beeender** [comment]
:shipit:
